From 3e4947df4ed3a4628b42307fcdb48e6d6054b990 Mon Sep 17 00:00:00 2001
From: Huang Chaofan <chaofan.huang@nxp.com>
Date: Fri, 22 Jun 2018 17:59:19 +0800
Subject: [PATCH 4063/5242] MLK-18664 VPU: Send eos event when receive
 VID_API_EVENT_FIRMWARE_XCPT

commit  4cb808eb6997117e6b9a432774cd346fe870df17 from
https://source.codeaurora.org/external/imx/linux-imx.git

Send eos event when receive VID_API_EVENT_FIRMWARE_XCPT to make vpu decoder
exit when firmware has error

Signed-off-by: Huang Chaofan <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index c86f7c5..098e8cd 100644
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -1651,7 +1651,7 @@ static void vpu_api_event_handler(struct vpu_ctx *ctx, u_int32 uStrIdx, u_int32
 					}
 				}
 				up(&This->drv_q_lock);
-				if (buffer_flag == false)
+				if (buffer_flag == false && !ctx->firmware_finished)
 					vpu_dbg(LVL_ERR, "error: don't find the right buffer for VID_API_CMD_FS_ALLOC\n");
 			} else if (ctx->wait_rst_done) {
 				u_int32 i;
@@ -1820,8 +1820,14 @@ static void vpu_api_event_handler(struct vpu_ctx *ctx, u_int32 uStrIdx, u_int32
 		v4l2_event_queue_fh(&ctx->fh, &ev); //notfiy app stream eos reached
 
 	}	break;
-	case VID_API_EVENT_FIRMWARE_XCPT:
+	case VID_API_EVENT_FIRMWARE_XCPT: {
+		const struct v4l2_event ev = {
+			.type = V4L2_EVENT_EOS
+		};
+		v4l2_event_queue_fh(&ctx->fh, &ev);
 		vpu_dbg(LVL_ERR, "warning: FIRMWARE hang, and send event VID_API_EVENT_FIRMWARE_XCPT\n");
+		}
+		break;
 	case VID_API_EVENT_DEC_CFG_INFO:
 		break;
 	default:
-- 
1.7.9.5

